### PR TITLE
New farm refactor

### PR DESCRIPTION
**Pull request by catalinnnn**
Copied farm to multi_token_farm. Might be useful later.
Updated farm to to a single token farm. Less setup, less code, more info on init

### CODE REVIEW

**Review by bogdan-rosianu — APPROVED** [review]

File: multi_token_farm/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,785 @@
...
+
+    #[payable("*")]
+    #[endpoint]
+    fn exitFarm(
+        &self,
+        #[payment_token] payment_token_id: TokenIdentifier,
+        #[payment] liquidity: Self::BigUint,
+    ) -> SCResult<MultiResult2<TokenAmountPair<Self::BigUint>, TokenAmountPair<Self::BigUint>>>
+    {
+        //require!(self.is_active(), "Not active");
```
Comment (sasurobert) [inline multi_token_farm/src/lib.rs:226]: please do not leave commented lines. If you have commented lines, add a TODO as well.

File: multi_token_farm/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,785 @@
...
+                );
+            }
+        }
+    }
+
+    #[view(calculateRewardsForGivenPosition)]
+    fn calculate_rewards_for_given_position(
+        &self,
+        token_nonce: u64,
+        liquidity: Self::BigUint,
```
Comment (sasurobert) [inline multi_token_farm/src/lib.rs:426]: change this function to get attributes.

File: multi_token_farm/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,785 @@
...
+        let new_farm_attributes = FarmTokenAttributes::<Self::BigUint> {
+            farmed_token_id: farm_attributes.farmed_token_id,
+            total_farmed_tokens: farmed_token_amount,
+            total_initial_worth: initial_worth,
+            total_amount_liquidity: re_added_liquidity.clone(),
+            entering_epoch: farm_attributes.entering_epoch,
+        };
+
+        // Create and send the new farm tokens.
+        let farm_tokens_to_create = &re_added_liquidity + &Self::BigUint::from(1u64);
```
Comment (sasurobert) [inline multi_token_farm/src/lib.rs:333]: no need to create +1.

File: elrond_dex_farm/src/liquidity_pool.rs [hunk 4]
```diff
@@ -3,100 +3,104 @@ elrond_wasm::derive_imports!();
...
     ) -> SCResult<Self::BigUint> {
-        let mut total_supply = self.total_supply().get();
+        require!(liquidity > &0, "Amount needs to be greater than 0");
+        require!(enter_amount > &0, "Amount needs to be greater than 0");
+
+        let reward_token_id = self.reward_token_id().get();
+        let farming_token_id = self.farming_token_id().get();
         let mut virtual_reserves = self.virtual_reserves().get();
+        let mut total_supply = self.total_supply().get();
+        require!(&total_supply > liquidity, "Not enough supply");
```
[the remark this answers is not in the document]
Comment (catalinnnn) [reply elrond_dex_farm/src/liquidity_pool.rs:56]: Yes. its just a sanity check

File: elrond_dex_farm/src/lib.rs [hunk 5]
```diff
@@ -363,92 +224,68 @@ pub trait Farm: liquidity_pool::LiquidityPoolModule + rewards::RewardsModule {
...
 
-        let reward = self.calculate_reward_for_given_liquidity(
+        let attributes = self.decode_attributes(&attributes_raw)?;
+        require!(
+            liquidity <= attributes.total_liquidity_amount,
+            "Bad arguments"
+        );
+
+        let entering_amount =
+            &attributes.total_entering_amount * &liquidity / attributes.total_liquidity_amount;
```
[the remark this answers is not in the document]
Comment (catalinnnn) [reply elrond_dex_farm/src/lib.rs:249]: No

File: elrond_dex_farm/src/lib.rs [hunk 6]
```diff
@@ -285,71 +159,58 @@ pub trait Farm: liquidity_pool::LiquidityPoolModule + rewards::RewardsModule {
...
-        let farmed_token_amount = &farm_attributes.total_farmed_tokens * &liquidity
-            / farm_attributes.total_amount_liquidity.clone();
-        require!(farmed_token_amount > 0, "Cannot unfarm with 0 farmed_token");
-        self.burn_tokens(&payment_token_id, token_nonce, &liquidity);
-
-        // Remove liquidity and send rewards. No penalty.
+        // Read the attributes from the received SFT.
+        let farm_attributes = self.get_farm_attributes(&payment_token_id, token_nonce)?;
+        let entering_amount = &farm_attributes.total_entering_amount * &liquidity
+            / farm_attributes.total_liquidity_amount.clone();
```
[the remark this answers is not in the document]
Comment (catalinnnn) [reply elrond_dex_farm/src/lib.rs:172]: No

File: elrond_dex_farm/src/liquidity_pool.rs [hunk 7]
```diff
@@ -3,100 +3,104 @@ elrond_wasm::derive_imports!();
...
-        farming_pool_token_id: TokenIdentifier,
-        farmed_token_id: TokenIdentifier,
-    ) -> SCResult<Self::BigUint> {
-        require!(amount > 0, "Amount needs to be greater than 0");
+pub trait LiquidityPoolModule: rewards::RewardsModule + config::ConfigModule {
+    fn add_liquidity(&self, amount: &Self::BigUint) -> SCResult<Self::BigUint> {
+        require!(amount > &0, "Amount needs to be greater than 0");
+        let reward_token_id = self.reward_token_id().get();
+        let farming_token_id = self.farming_token_id().get();
+        let mut total_supply = self.total_supply().get();
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply elrond_dex_farm/src/liquidity_pool.rs:15]: virtual reserve and actual reserve is needed. As we mint tokens only on claimRewards / exitFarm.

**Pull request merged**